### PR TITLE
Enumerable methods project complete

### PR DESCRIPTION
**Pull request by fivan18**
# Completed own Enumerable module

This project consisted of implementing my own version of the next Ruby’s enumerable methods.

- [ ] each
- [ ] each_with_index
- [ ] select
- [ ] all?
- [ ] any?
- [ ] none?
- [ ] count
- [ ] map
- [ ] inject

## Authors

👤 **Ivan Ulises Guzman Sanchez**
- Github: [@fivan18](https://github.com/fivan18)

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job :+1: 
You have met most requirements for the project. Your commits are descriptive and the code looks clean.
However, we can not merge this PR yet. 
- [x] make your README more descriptive.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
- [x] read my comments for more suggestions

### Martin Afani

File: lib/enumerable_1.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    # return an enumerator
```
Comment (whiz25) [inline lib/enumerable_1.rb]: remove this comment and any other you have

File: lib/enumerable_1.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                  end
+                else
+                  proc { |item, _patt| item }
+                end
+
+    my_each do |item|
+      return false unless condition.call(item, pattern)
+    end
+    true
+  end
```
Comment (whiz25) [inline lib/enumerable_1.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: lib/enumerable_1.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                    end
+                  end
+                else
+                  proc { |item, _patt| item }
+                end
+    my_each do |item|
+      return true if condition.call(item, pattern)
+    end
+    false
+  end
```
Comment (whiz25) [inline lib/enumerable_1.rb]: - [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: lib/enumerable_1.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                else
+                  proc { |item, _patt| item }
+                end
+
+    my_each do |item|
+      return false if condition.call(item, pattern)
+    end
+    true
+  end
+end
```
Comment (whiz25) [inline lib/enumerable_1.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: lib/enumerable_2.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      my_each_with_index do |item, index|
+        initial = args.empty? && index.zero? ? initial : yield(initial, item)
+      end
+    else
+      my_each_with_index do |item, index|
+        initial = args.size == 1 && index.zero? ? initial : initial.send(sym, item)
+      end
+    end
+    initial
+  end
```
Comment (whiz25) [inline lib/enumerable_2.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

Comment (fivan18) [reply lib/enumerable_2.rb]: https://github.com/fivan18/enumerables/blob/77f2de9dc296c28b1e9e7763667b155d0df44fcb/spec/enumerable_spec.rb#L142-L147

Thanks for the code review @whiz25,

These are my test cases that are related with your comments, they passed. Let me know yours please to understand what you mean.

Comment (wathigo) [inline lib/enumerable_2.rb]: **[REQUIRED]**
* Combines all elements of enum by applying a binary operation, specified by a block.
Screenshot from 2020-01-06 01-40-02

* When a symbol is specified combines each element of the collection by applying the symbol as a named method.
Screenshot from 2020-01-06 01-42-45

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required 
### Great work so far :clap:  
We have only two failing tests coming from `my_inject` method.
See the comments :point_down: 

**After implementing the changes, submit another request for review.**

**Happy Coding :computer: 
Simon Wathigo**

**Review by whiz25 — APPROVED** [review]
## Hey @fivan18 
It's a :+1: from me. Merge and proceed.

### Martin Afani

**Pull request merged**